lib/util/pre_receive_util.js

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright (c) 2016, Two Sigma Open Source
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of git-meta nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+"use strict";
+
+/**
+ * This module contains the pre-recieve hook for a git-meta server.  Only
+ the meta-repo should use this hook.
+ */
+
+const NodeGit = require("nodegit");
+const GitUtil = require("./git_util");
+const async = require("generator-async");
+const split = require("split");
+const co = require("co");
+const path = require("path");
+const assert  = require("chai").assert;
+
+const NOTES_REF = "refs/notes/git-meta/subrepo-check";
+
+function identity(v) {
+    return v;
+}
+
+function flatmap(arr, fn) {
+    return arr.map(fn).filter(identity);
+}
+
+function *getParentTrees(parentEntries) {
+    const entryParentTreeEntries = parentEntries.filter(function isTree(entry) {
+        return entry.isTree();
+    });
+    return yield entryParentTreeEntries.map(
+        function *(entry) {
+            return yield entry.getTree();
+        });
+}
+
+class PreReceiveCheck {
+
+    *urlToLocalPath(repo, url) {
```

Comment:
Unfortunately, this is one of the suckier parts of JavaScript. You can get something close to private with ES6 symbols: http://stackoverflow.com/questions/22156326/private-properties-in-javascript-es6-classes, or you can do it the "old school" way with closures as I did with the `SubmoduleRebaser` in `lib/util/rebase.js`.